File: dist_build/_staging/rag_core/embeddings.py
"""
rag_core.embeddings — Embedding Model Manager
================================================

Handles loading, fallback, and caching of embedding models.
Supports code-specific models (GraphCodeBERT) and general models (MiniLM).
"""

from __future__ import annotations

import logging
from pathlib import Path
from typing import List, Optional, Tuple

import numpy as np

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Model registry — best → fastest
# ---------------------------------------------------------------------------

EMBEDDING_MODELS: List[Tuple[str, str, int]] = [
    # Code-specific (best for code/function similarity)
    ("microsoft/graphcodebert-base", "code", 768),
    ("microsoft/codebert-base", "code", 768),
    # General purpose (best for natural language documents)
    ("all-MiniLM-L6-v2", "general", 384),
    ("paraphrase-MiniLM-L6-v2", "general", 384),
    ("paraphrase-multilingual-MiniLM-L12-v2", "multilingual", 384),
]

# Cache dir for downloaded models
_DEFAULT_CACHE = Path.home() / ".cache" / "rag_core" / "models"


class EmbeddingManager:
    """
    Manages loading and inference of embedding models.

    Features:
    - Priority-ordered model hierarchy with automatic fallback
    - GPU acceleration with FP16 when available
    - Batch encoding with progress callbacks
    - Normalised embeddings for cosine similarity via dot product
    """

    def __init__(
        self,
        model_name: Optional[str] = None,
        prefer_code: bool = False,
        device: Optional[str] = None,
        use_fp16: bool = True,
    ):
        self.model_name = model_name
        self.prefer_code = prefer_code
        self._device = device
        self._use_fp16 = use_fp16
        self._model = None
        self._model_type: str = "none"
        self._dim: int = 0

    @property
    def is_loaded(self) -> bool:
        return self._model is not None

    @property
    def model_type(self) -> str:
        return self._model_type

    @property
    def dimension(self) -> int:
        return self._dim

    def load(self) -> bool:
        """Load the best available embedding model.

        Returns True if a model was loaded, False if no model available.
        """
        try:
            from sentence_transformers import SentenceTransformer
        except ImportError:
            logger.warning("sentence-transformers not installed")
            return False

        models_to_try = []
        if self.model_name:
            models_to_try.append((self.model_name, "custom", 768))
        if self.prefer_code:
            models_to_try.extend(EMBEDDING_MODELS)
        else:
            # Put general models first
            models_to_try.extend(
                [m for m in EMBEDDING_MODELS if m[1] != "code"] + [m for m in EMBEDDING_MODELS if m[1] == "code"]
            )

        for name, mtype, dim in models_to_try:
            try:
                logger.info("Loading embedding model: %s (%s)", name, mtype)
                device = self._device
                if device is None:
                    import torch

                    device = "cuda" if torch.cuda.is_available() else "cpu"

                model = SentenceTransformer(name, device=device)

                # Enable FP16 on GPU
                if self._use_fp16 and device == "cuda":
                    model = model.half()

                self._model = model
                self._model_type = mtype
                self._dim = model.get_sentence_embedding_dimension()
                logger.info(
                    "Loaded %s — dim=%d, type=%s, device=%s",
                    name,
                    self._dim,
                    mtype,
                    device,
                )
                return True
            except Exception as e:
                logger.debug("Failed to load %s: %s", name, e)

        return False

    def encode(
        self,
        texts: List[str],
        *,
        batch_size: int = 32,
        normalize: bool = True,
        show_progress: bool = False,
    ) -> np.ndarray:
        """Encode texts into embeddings.

        Args:
            texts: List of texts to encode.
            batch_size: Batch size for encoding.
            normalize: Whether to L2-normalise (enables dot-product = cosine).
            show_progress: Show tqdm progress bar.

        Returns:
            numpy array of shape (len(texts), dim).
        """
        if self._model is None:
            raise RuntimeError("No embedding model loaded. Call load() first.")

        return self._model.encode(
            texts,
            batch_size=batch_size,
            normalize_embeddings=normalize,
            show_progress_bar=show_progress,
        )

    def encode_single(self, text: str, normalize: bool = True) -> np.ndarray:
        """Encode a single text into an embedding vector."""
        return self.encode([text], normalize=normalize)[0]

    def cosine_similarity(self, query_vec: np.ndarray, corpus_vecs: np.ndarray) -> np.ndarray:
        """Compute cosine similarity between query and corpus.

        Assumes vectors are L2-normalised (dot product = cosine).
        """
        return corpus_vecs @ query_vec

    def save_embeddings(self, embeddings: np.ndarray, path: Path) -> None:
        """Save embeddings to disk for persistent caching."""
        path.parent.mkdir(parents=True, exist_ok=True)
        np.save(str(path), embeddings)
        logger.info("Saved embeddings to %s (%s)", path, embeddings.shape)

    def load_embeddings(self, path: Path) -> Optional[np.ndarray]:
        """Load cached embeddings from disk."""
        if not path.exists():
            return None
        try:
            emb = np.load(str(path))
            logger.info("Loaded embeddings from %s (%s)", path, emb.shape)
            return emb
        except Exception as e:
            logger.warning("Failed to load embeddings from %s: %s", path, e)
            return None

File: rust_output/src/embeddings.rs
/// rag_core.embeddings — Embedding Model Manager
/// ================================================
/// 
/// Handles loading, fallback, and caching of embedding models.
/// Supports code-specific models (GraphCodeBERT) and general models (MiniLM).

use anyhow::{Result, Context};
use std::path::PathBuf;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static EMBEDDING_MODELS: std::sync::LazyLock<Vec<(String, String, i64)>> = std::sync::LazyLock::new(|| Vec::new());

pub const _DEFAULT_CACHE: &str = "Path.home() / '.cache' / 'rag_core' / 'models";

/// Manages loading and inference of embedding models.
/// 
/// Features:
/// - Priority-ordered model hierarchy with automatic fallback
/// - GPU acceleration with FP16 when available
/// - Batch encoding with progress callbacks
/// - Normalised embeddings for cosine similarity via dot product
#[derive(Debug, Clone)]
pub struct EmbeddingManager {
    pub model_name: String,
    pub prefer_code: String,
    pub _device: String,
    pub _use_fp16: String,
    pub _model: Option<serde_json::Value>,
    pub _model_type: String,
    pub _dim: i64,
}

impl EmbeddingManager {
    pub fn new(model_name: Option<String>, prefer_code: bool, device: Option<String>, use_fp16: bool) -> Self {
        Self {
            model_name,
            prefer_code,
            _device: device,
            _use_fp16: use_fp16,
            _model: None,
            _model_type: String::new(),
            _dim: 0,
        }
    }
    pub fn is_loaded(&self) -> bool {
        self._model.is_some()
    }
    pub fn model_type(&self) -> &String {
        self._model_type
    }
    pub fn dimension(&self) -> i64 {
        self._dim
    }
    /// Load the best available embedding model.
    /// 
    /// Returns true if a model was loaded, false if no model available.
    pub fn load(&mut self) -> Result<bool> {
        // Load the best available embedding model.
        // 
        // Returns true if a model was loaded, false if no model available.
        // try:
        {
            // TODO: from sentence_transformers import SentenceTransformer
        }
        // except ImportError as _e:
        let mut models_to_try = vec![];
        if self.model_name {
            models_to_try.push((self.model_name, "custom".to_string(), 768));
        }
        if self.prefer_code {
            models_to_try.extend(EMBEDDING_MODELS);
        } else {
            models_to_try.extend((EMBEDDING_MODELS.iter().filter(|m| m[1] != "code".to_string()).map(|m| m).collect::<Vec<_>>() + EMBEDDING_MODELS.iter().filter(|m| m[1] == "code".to_string()).map(|m| m).collect::<Vec<_>>()));
        }
        for (name, mtype, dim) in models_to_try.iter() {
            // try:
            {
                logger.info("Loading embedding model: %s (%s)".to_string(), name, mtype);
                let mut device = self._device;
                if device.is_none() {
                    // TODO: import torch
                    let mut device = if torch.cuda.is_available() { "cuda".to_string() } else { "cpu".to_string() };
                }
                let mut model = SentenceTransformer(name, /* device= */ device);
                if (self._use_fp16 && device == "cuda".to_string()) {
                    let mut model = model.half();
                }
                self._model = model;
                self._model_type = mtype;
                self._dim = model.get_sentence_embedding_dimension();
                logger.info("Loaded %s — dim=%d, type=%s, device=%s".to_string(), name, self._dim, mtype, device);
                true
            }
            // except Exception as e:
        }
        Ok(false)
    }
    /// Encode texts into embeddings.
    /// 
    /// Args:
    /// texts: List of texts to encode.
    /// batch_size: Batch size for encoding.
    /// normalize: Whether to L2-normalise (enables dot-product = cosine).
    /// show_progress: Show tqdm progress bar.
    /// 
    /// Returns:
    /// numpy array of shape (len(texts), dim).
    pub fn encode(&mut self, texts: Vec<String>) -> Result<np::ndarray> {
        // Encode texts into embeddings.
        // 
        // Args:
        // texts: List of texts to encode.
        // batch_size: Batch size for encoding.
        // normalize: Whether to L2-normalise (enables dot-product = cosine).
        // show_progress: Show tqdm progress bar.
        // 
        // Returns:
        // numpy array of shape (len(texts), dim).
        if self._model.is_none() {
            return Err(anyhow::anyhow!("RuntimeError('No embedding model loaded. Call load() first.')"));
        }
        Ok(self._model.encode(texts, /* batch_size= */ batch_size, /* normalize_embeddings= */ normalize, /* show_progress_bar= */ show_progress))
    }
    /// Encode a single text into an embedding vector.
    pub fn encode_single(&mut self, text: String, normalize: bool) -> np::ndarray {
        // Encode a single text into an embedding vector.
        self.encode(vec![text], /* normalize= */ normalize)[0]
    }
    /// Compute cosine similarity between query and corpus.
    /// 
    /// Assumes vectors are L2-normalised (dot product = cosine).
    pub fn cosine_similarity(&self, query_vec: np::ndarray, corpus_vecs: np::ndarray) -> np::ndarray {
        // Compute cosine similarity between query and corpus.
        // 
        // Assumes vectors are L2-normalised (dot product = cosine).
        (corpus_vecs /* op */ query_vec)
    }
    /// Save embeddings to disk for persistent caching.
    pub fn save_embeddings(&self, embeddings: np::ndarray, path: PathBuf) -> () {
        // Save embeddings to disk for persistent caching.
        path.parent().unwrap_or(std::path::Path::new("")).create_dir_all();
        numpy.save(path.to_string(), embeddings);
        logger.info("Saved embeddings to %s (%s)".to_string(), path, embeddings::shape);
    }
    /// Load cached embeddings from disk.
    pub fn load_embeddings(&self, path: PathBuf) -> Result<Option<np::ndarray>> {
        // Load cached embeddings from disk.
        if !path.exists() {
            None
        }
        // try:
        {
            let mut emb = numpy.load(path.to_string());
            logger.info("Loaded embeddings from %s (%s)".to_string(), path, emb.shape);
            emb
        }
        // except Exception as e:
    }
}
